implement LogitNormal univariate distribution

bgctw opened the pull request:
I started an implementation of the LogitNormal distribution by adapting lognormal.jl, which provides the pdf and cumulative functionality. There is no analytical solution for the moments of the LogitNormal, but I intend to add numerical approximations in future.

#850

matbesancon commented:
Sorry about the delay on this. I think we should not make Distributions depend on more external packages. (the goal would be to remove more than we add at this point). I would be ok to implement `LogitNormal` and leave all non-analytical functions as error. The other solution would be to build a package with a `LogitNormal` type which implements the `Distributions` interface

bgctw commented:
I removed the numerical code that depends on Optim, and moved those functions to a future package [NormalTransforms](https://git.bgc-jena.mpg.de/twutz/NormalTransforms.jl).

In this way, a broader community can benefit from the partial implementation of the logitnormal distribution with default availability with Distributions.jl. For many use cases,  the mode and the quantiles are sufficient.

matbesancon commented on test/logitnormal.jl:
```diff
@@ -0,0 +1,70 @@
+using StatsFuns
+
+
+function tmpF()
```
is this needed?

bgctw replied:
I removed the `tmpF()` where I stored interactive code.
The usage of `StatsFuns` is necessary for the logistic function used at line 21.

matbesancon commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -0,0 +1,174 @@
+end
+
+median(d::LogitNormal) = logistic(d.μ)
+function mode(d::LogitNormal)
```
I don't think we should even implement them here, that way it can be implemented in a new package. Otherwise which method is used between this one and the new "real" method is not deterministic

```julia
function mean(d::LogitNormal; kwargs...) 
    # estimateMean(d,kwargs...) # moved to NormalTransforms
    @warn "Mean of logitnomral currently not implemented. Return missing"
    missing
end 
```

bgctw replied:
This means, The recommended way is to remove those methods completely? 
And maybe leave a comment in the code?

matbesancon commented:
Regarding my last comment, this is not the expected behaviour of missing (used for missing data, not missing implementations). Better having a `MethodError`, which is what will happen if it is not implemented

matbesancon commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -0,0 +1,153 @@
+# stdlogitx(d::LogitNormal) = d.σ
+
+# mean, mode, and variance 
+# moved to NormalTrnasforms because
```
```suggestion
# moved to NormalTransforms because
```

matbesancon commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -0,0 +1,153 @@
+
+# mean, mode, and variance 
+# moved to NormalTrnasforms because
+# they depend on Optim package.
```
```suggestion
# they depend on the Optim package.
```

matbesancon commented:
looks good, merging in 12 hours if no objection

bgctw commented:
Thanks, I have no objections.

matbesancon commented:
@bgctw totally forgot, could you add the distribution in the corresponding bloc in `docs/src/univariate.md`?

matbesancon commented:
@bgctw I was thinking if some maintainers have no objection, but thanks :smile:

mschauer commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -0,0 +1,153 @@
+"""
+    LogitNormal(μ,σ)
+
+The *logit normal distribution* is the distribution of the logistic function 
```
bad wording, it is not the distribution of a function `f` but the distribution of a transformed variable `f(X)`

mschauer commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -0,0 +1,153 @@
+#TODO check pd and logpdf
+function pdf(d::LogitNormal{T}, x::Real) where T<:Real 
+    if zero(x) < x < one(x) 
+        return normpdf(d.μ, d.σ, logit(x)) / x / (1-x)
```
Does it numerically make sense to use `a/(b*c)` instead of `a/b/c`?

matbesancon replied:
a/b/c always bugs me, never sure in which order to take it

mschauer commented on src/univariate/continuous/logitnormal.jl:
```diff
@@ -131,7 +136,7 @@ invlogccdf(d::LogitNormal, lq::Real) = logistic(norminvlogccdf(d.μ, d.σ, lq))
     #TODO check
     (μ, σ) = params(d)
-    0 < x < 1 ? - ((log(x) - μ) / (σ^2) + 1) / x /(1-x) : zero(T)
+    0 < x < 1 ? - ((log(x) - μ) / ((σ^2) + 1) * x * (1-x)) : zero(T)
```
Huh?

matbesancon commented:
thanks @bgctw!

mschauer commented:
@matbesancon Is there a change to runtests.jl missing?

matbesancon commented:
oh boy, yes absolutely
